### PR TITLE
[SPARK-6405] Limiting the maximum Kryo buffer size to be 2GB.

### PR DESCRIPTION
**Pull request by mccheah**
Kryo buffers are backed by byte arrays, but primitive arrays can only be
up to 2GB in size. It is misleading to allow users to set buffers past
this size.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala [hunk 1]
```diff
@@ -49,10 +49,20 @@ class KryoSerializer(conf: SparkConf)
   with Logging
   with Serializable {
 
-  private val bufferSize =
-    (conf.getDouble("spark.kryoserializer.buffer.mb", 0.064) * 1024 * 1024).toInt
+  private val bufferSizeMb: Double = conf.getDouble("spark.kryoserializer.buffer.mb", 0.064)
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: Declaring the type here *is* more descriptive, but I'd leave it out to be consistent with the other variables in the class.

**Comment by liancheng** [comment]
LGTM

File: core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala [hunk 2]
```diff
@@ -49,10 +49,20 @@ class KryoSerializer(conf: SparkConf)
...
+      s"2048 mb, got: + $bufferSizeMb mb.")
+  }
+  private val bufferSize = (bufferSizeMb * 1024 * 1024).toInt
+
+  val maxBufferSizeMb = conf.getInt("spark.kryoserializer.buffer.max.mb", 64)
+  if (maxBufferSizeMb >= 2048) {
+    throw new IllegalArgumentException("spark.kryoserializer.buffer.max.mb must be less than " +
+      s"2048 mb, got: + $maxBufferSizeMb mb.")
+  }
+  private val maxBufferSize = maxBufferSizeMb * 1026 * 1024
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: Is this meant to be *1026* * 1024?

Comment (mccheah) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: good catch =D

**Comment by sryza** [comment]
LGTM pending jenkins

**Comment by pwendell** [comment]
LGTM - merged into master.

**Pull request closed without merging**